applications/tari_base_node/src/recovery.rs

```diff
@@ -0,0 +1,166 @@
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+use log::*;
+use std::{
+    io::{self, Write},
+    path::Path,
+    sync::Arc,
+};
+use tari_app_utilities::utilities::ExitCodes;
+use tari_common::{DatabaseType, GlobalConfig};
+use tari_core::{
+    chain_storage::{
+        create_lmdb_database,
+        create_recovery_lmdb_database,
+        remove_lmdb_database,
+        BlockchainBackend,
+        BlockchainDatabase,
+        BlockchainDatabaseConfig,
+        Validators,
+    },
+    consensus::{ConsensusManagerBuilder, Network as NetworkType},
+    transactions::types::CryptoFactories,
+    validation::{
+        block_validators::{FullConsensusValidator, StatelessBlockValidator},
+        mocks::MockValidator,
+    },
+};
+use tari_mmr::MmrCacheConfig;
+pub const LOG_TARGET: &str = "base_node::app";
+
+pub fn initiate_recover_db(node_config: &GlobalConfig) -> Result<(), ExitCodes> {
+    // create recovery db
+    match &node_config.db_type {
+        DatabaseType::LMDB(p) => {
+            let _backend = create_recovery_lmdb_database(&p).map_err(|err| {
+                error!(target: LOG_TARGET, "{}", err);
+                ExitCodes::UnknownError
+            })?;
+        },
+        _ => {
+            error!(target: LOG_TARGET, "Recovery mode is only available for LMDB");
+            return Err(ExitCodes::UnknownError);
+        },
+    };
+    Ok(())
+}
+
+pub async fn run_recovery(node_config: &GlobalConfig) -> Result<(), String> {
+    println!("string recovery mode");
+    let (temp_db, main_db) = match &node_config.db_type {
+        DatabaseType::LMDB(p) => {
+            let backend =
+                create_lmdb_database(&p, node_config.db_config.clone(), MmrCacheConfig::default()).map_err(|e| {
+                    error!(target: LOG_TARGET, "Error opening db: {}", e);
+                    format!("Could not open DB: {}", e.to_string())
+                })?;
+            let new_path = Path::new(&p).join("temp_recovery");
+
+            let temp = create_lmdb_database(&new_path, node_config.db_config.clone(), MmrCacheConfig::default())
+                .map_err(|e| {
+                    error!(target: LOG_TARGET, "Error opening recovery db: {}", e);
+                    format!("Could not open recovery DB: {}", e.to_string())
```

Comment:
The `{}` formatter calls the same code as `to_string`

```suggestion
                    anyhow!("Could not open recovery DB: {}", e)
```